internal/controller/pkg/signature/reconciler.go

```diff
@@ -0,0 +1,435 @@
+}
+
+// SetupFunction adds a controller that reconciles Functions.
+func SetupFunction(mgr ctrl.Manager, o controller.Options) error {
+	n := "package-signature-verification/" + strings.ToLower(v1.FunctionGroupKind)
+	np := func() v1.Package { return &v1.Function{} }
+
+	clientset, err := kubernetes.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		return errors.Wrap(err, errNewKubernetesClient)
+	}
+
+	cosignValidator, err := NewCosignValidator(mgr.GetClient(), o.Namespace)
+	if err != nil {
+		return errors.Wrap(err, "cannot create cosign validator")
+	}
+
+	log := o.Logger.WithValues("controller", n)
+	cb := ctrl.NewControllerManagedBy(mgr).
+		Named(n).
+		For(&v1.Function{}).
+		Watches(&v1beta1.ImageConfig{}, enqueueFunctionsForImageConfig(mgr.GetClient(), log))
+
+	ro := []ReconcilerOption{
+		WithNewPackageFn(np),
+		WithNamespace(o.Namespace),
+		WithServiceAccount(o.ServiceAccount),
+		WithDefaultRegistry(o.DefaultRegistry),
+		WithConfigStore(xpkg.NewImageConfigStore(mgr.GetClient(), o.Namespace)),
+		WithValidator(cosignValidator),
+		WithLogger(log),
+	}
+
+	return cb.WithOptions(o.ForControllerRuntime()).
+		Complete(ratelimiter.NewReconciler(n, errors.WithSilentRequeueOnConflict(NewReconciler(mgr.GetClient(), clientset, ro...)), o.GlobalRateLimiter))
+}
+
+// NewReconciler creates a new package reconciler for signature verification.
+func NewReconciler(client client.Client, clientset kubernetes.Interface, opts ...ReconcilerOption) *Reconciler {
+	r := &Reconciler{
+		client:    client,
+		clientset: clientset,
+		log:       logging.NewNopLogger(),
+	}
+
+	for _, f := range opts {
+		f(r)
+	}
+
+	return r
+}
+
+// Reconcile packages and verify signatures if configured.
+func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(ctx, reconcileTimeout)
+	defer cancel()
+
+	p := r.newPackage()
+	if err := r.client.Get(ctx, req.NamespacedName, p); err != nil {
+		if kerrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+
+		log.Debug(errGetPackage, "error", err)
+		p.SetConditions(v1.VerificationIncomplete(errors.Wrap(err, errGetPackage)))
+		_ = r.client.Status().Update(ctx, p)
+		return reconcile.Result{}, errors.Wrap(err, errGetPackage)
+	}
+
+	log = log.WithValues(
+		"uid", p.GetUID(),
+		"version", p.GetResourceVersion(),
+		"name", p.GetName(),
+	)
+
+	// If signature verification is already complete, nothing to do here.
+	// A package is deployed once signature verification is complete which means
+	// either the verification skipped or succeeded. Once we have this condition,
+	// it doesn't make sense to verify the signature again since the package is
+	// already deployed.
+	if cond := p.GetCondition(v1.TypeSignatureVerificationComplete); cond.Status == corev1.ConditionTrue {
+		return reconcile.Result{}, nil
+	}
```

Comment:
A package's source is mutable though, right?

For example is it possible we validated `xpkg.upbound.io/negz/function-sometimes-evil:v0.1.0`, but then the source is changed to `xpkg.upbound.io/negz/function-sometimes-evil:v0.2.0` which wouldn't pass validation?

Would it make more sense for the signature controller to operate on package revisions, which have immutable sources?

Typically the package manager controller is only going to make a HEAD request resolve the tag to a digest in order to create a revision (though iirc it does fall back to pulling the whole package sometimes). It's presumably the revision controller where we need to make sure it doesn't proceed to install the package until it's verified.

Reply:
Done, fixing e2e's now ...

Reply:
Ah, remembered another reason why I didn't like it to be on revisions and noticed another nit(?) that makes things a little weird.

I executed e2e test steps manually as follows:

```bash
$ cd test/e2e/manifests/pkg/image-config/signature-verification/with-key
$ kubectl apply -f image-config.yaml
$ kubectl apply -f configuration-unsigned.yaml

$ kubectl get configurations
NAME                                INSTALLED   HEALTHY   PACKAGE                                                               AGE
e2e-configuration-signed-with-key   True        False     xpkg.upbound.io/crossplane/e2e-configuration-signed-with-key:v0.1.0   24s
$ kubectl get configurationrevisions
NAME                                             HEALTHY   REVISION   IMAGE                                                                 STATE    DEP-FOUND   DEP-INSTALLED   AGE
e2e-configuration-signed-with-key-e0adba255c20   False     1          xpkg.upbound.io/crossplane/e2e-configuration-signed-with-key:v0.1.0   Active                               43s
$ kubectl get configurationrevision ... -o yaml
...
status:
  conditions:
  - lastTransitionTime: "2024-10-28T19:20:09Z"
    reason: AwaitingSignatureVerification
    status: "False"
    type: Healthy
  - lastTransitionTime: "2024-10-28T19:20:12Z"
    message: 'Signature verification failed using ImageConfig named "e2e-signed-with-key":
      authority "verify e2e signed with key": signature verification failed with no
      signatures found'
    reason: SignatureVerificationFailed
    status: "False"
    type: Verified
```

Configuration says "Installed", but is it really? Probably, already the case if any step in the revision reconciler fails. This is the "nit" one.

Continue by applying the signed one:

```
$ kubectl apply -f configuration-signed.yaml
$ kubectl get configurations
NAME                                INSTALLED   HEALTHY   PACKAGE                                                               AGE
e2e-configuration-signed-with-key   True        True      xpkg.upbound.io/crossplane/e2e-configuration-signed-with-key:v0.2.0   4m1s
$ kubectl get configurationrevisions.pkg.crossplane.io
NAME                                             HEALTHY   REVISION   IMAGE                                                                 STATE      DEP-FOUND   DEP-INSTALLED   AGE
e2e-configuration-signed-with-key-1765fb139d01   True      2          xpkg.upbound.io/crossplane/e2e-configuration-signed-with-key:v0.2.0   Active                                 88s
e2e-configuration-signed-with-key-e0adba255c20   False     1          xpkg.upbound.io/crossplane/e2e-configuration-signed-with-key:v0.1.0   Inactive                               5m10s
$ kubectl get configurationrevision ... -o yaml
...
status:
  conditions:
  - lastTransitionTime: "2024-10-28T19:23:57Z"
    reason: HealthyPackageRevision
    status: "True"
    type: Healthy
  - lastTransitionTime: "2024-10-28T19:23:55Z"
    message: Signature verification succeeded using ImageConfig named "e2e-signed-with-key"
    reason: SignatureVerificationSucceeded
    status: "True"
    type: Verified
```

All looks good, except the first revision, which was failing verification, will keep retrying the signature verification unnecessarily (?). Is this ok already? or should I only retry a failing verification for latest revision? or active one? 🤔

Reply:
> Configuration says "Installed", but is it really?

The problem is that the package manager says the package is installed even if there's only one revision that's never been healthy, right? As you say, I think it's pre-existing.

Looking at the manager Reconciler code it seems like:

* The package's Healthy condition always reflects the Healthy condition of the active revision
* The package's Installed condition is always true if there's an active revision

I could see an argument that this is the most conservative thing the manager controller could do. Once it's created a package revision it doesn't really know what changes (if any) the revision made to the system. For example an unhealthy revision could still have partially installed the package - e.g. creating CRDs, creating a deployment. So in that case I think it makes sense to say the package is installed, but unhealthy.

It is pretty nuanced and potentially confusing though. In this case there's only one revision and it  became unhealthy very early in its lifecycle (i.e. because it didn't pass validation) so nothing was really installed.

On the other hand, if another revision was created successfully before this one (that fails validation) then I'd argue the package was installed because it presumably created its Compositions, XRDs, etc and they should still exist despite the revision that created them being inactive (due to the owner refs to the package).

I feel like we should just leave this for the time being and see whether folks find it confusing in practice.

Reply:
> should I only retry a failing verification for latest revision? or active one?

Intuitively it makes sense to me to only attempt validation in the first place for the active revision. I don't think an inactive revision will ever actually try to pull an OCI image, right?

Reply:
Thanks for the response, updated accordingly.